node/core/av-store/src/lib.rs

```diff
@@ -19,52 +19,290 @@
+const KEEP_FINALIZED_CHUNK_FOR: u64 = 25 * 60 * 60;
+
+/// At which point in time we need to wakeup and do next pruning of blocks.
+/// Essenially this is the first element in the sorted array of pruning data,
+/// we just want to cache it here to avoid lifting the whole array just to look at the head.
+///
+/// This record exists under `NEXT_POV_PRUNING` key, if it does not either:
+///  a) There are no records and nothing has to be pruned.
+///  b) There are records but all of them are in `Included` state and do not have exact time to
+///     be pruned.
+#[derive(Decode, Encode)]
+struct NextPoVPruning(u64);
+
+impl NextPoVPruning {
+	// After which amount of seconds into the future from `now` this should fire.
+	fn should_fire_in(&self) -> Result<u64, Error> {
+		Ok(self.0.saturating_sub(SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs()))
+	}
+}
+
+/// At which point in time we need to wakeup and do next pruning of chunks.
+/// Essenially this is the first element in the sorted array of pruning data,
+/// we just want to cache it here to avoid lifting the whole array just to look at the head.
+///
+/// This record exists under `NEXT_CHUNK_PRUNING` key, if it does not either:
+///  a) There are no records and nothing has to be pruned.
+///  b) There are records but all of them are in `Included` state and do not have exact time to
+///     be pruned.
+#[derive(Decode, Encode)]
+struct NextChunkPruning(u64);
+
+impl NextChunkPruning {
+	// After which amount of seconds into the future from `now` this should fire.
+	fn should_fire_in(&self) -> Result<u64, Error> {
+		Ok(self.0.saturating_sub(SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs()))
+	}
+}
+
+/// Struct holding pruning timing configuration.
+/// The only purpose of this structure is to use different timing
+/// configurations in production and in testing.
+#[derive(Clone)]
+struct PruningConfig {
+	/// How long should a stored block stay available.
+	keep_stored_block_for: u64,
+
+	/// How long should a finalized block stay available.
+	keep_finalized_block_for: u64,
+
+	/// How long should a chunk of a finalized block stay available.
+	keep_finalized_chunk_for: u64
+}
+
+impl Default for PruningConfig {
+    fn default() -> Self {
+		Self {
+			keep_stored_block_for: KEEP_STORED_BLOCK_FOR,
+			keep_finalized_block_for: KEEP_FINALIZED_BLOCK_FOR,
+			keep_finalized_chunk_for: KEEP_FINALIZED_CHUNK_FOR,
+		}
+    }
+}
+
+#[derive(Debug, Decode, Encode, Eq, PartialEq)]
+enum CandidateState {
+	Stored,
+	Included,
+	Finalized,
+}
+
+#[derive(Debug, Decode, Encode, Eq)]
+struct PoVPruningRecord {
+	candidate_hash: Hash,
+	block_number: BlockNumber,
+	candidate_state: CandidateState,
+	prune_at: u64,
+}
+
+impl PartialEq for PoVPruningRecord {
+	fn eq(&self, other: &Self) -> bool {
+		self.candidate_hash == other.candidate_hash
+	}
+}
+
+impl Ord for PoVPruningRecord {
+	fn cmp(&self, other: &Self) -> Ordering {
+		self.prune_at.cmp(&other.prune_at)
+	}
+}
```

Comment:
`eq` might not always correspond to `Ordering::Equal` under this schema; [that's not supposed to happen](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html#how-can-i-implement-partialeq). I don't know what precisely the consequences of a mismatch would be, but my instinct is to trust the std docs and implement them such that they always match.

Reply:
The idea here is that `x <= y` and `x == y && x < y ` should not behave differently, which is kind of an invariant assumption when reading code (generally speaking).

Reply:
yes, i've added the check that i think fixes this